StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/MemberOrderHelper.cs

```diff
@@ -9,22 +10,72 @@
     /// <summary>
     /// Helper for dealing with member priority.
     /// </summary>
-    public struct MemberOrderHelper
+    internal struct MemberOrderHelper
     {
+        private static readonly ImmutableArray<SyntaxKind> TypeMemberOrder = ImmutableArray.Create(
+            SyntaxKind.ClassDeclaration,
+            SyntaxKind.StructDeclaration,
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.OperatorDeclaration,
+            SyntaxKind.ConversionOperatorDeclaration,
+            SyntaxKind.IndexerDeclaration,
+            SyntaxKind.PropertyDeclaration,
+            SyntaxKind.InterfaceDeclaration,
+            SyntaxKind.EnumDeclaration,
+            SyntaxKind.EventDeclaration,
+            SyntaxKind.DelegateDeclaration,
+            SyntaxKind.DestructorDeclaration,
+            SyntaxKind.ConstructorDeclaration,
+            SyntaxKind.FieldDeclaration,
+            SyntaxKind.NamespaceDeclaration);
+
         private readonly ModifierFlags modifierFlags;
+        private readonly int elementPriority;
         private readonly AccessLevel accessibilty;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="MemberOrderHelper"/> struct.
         /// </summary>
         /// <param name="member">The member to wrap.</param>
-        public MemberOrderHelper(MemberDeclarationSyntax member)
+        /// <param name="checks">The element ordering checks.</param>
+        internal MemberOrderHelper(MemberDeclarationSyntax member, ElementOrderingChecks checks)
         {
             this.Member = member;
             var modifiers = member.GetModifiers();
+            var type = member.Kind();
+            type = type == SyntaxKind.EventFieldDeclaration ? SyntaxKind.EventDeclaration : type;
 
-            this.modifierFlags = GetModifierFlags(modifiers);
-            this.accessibilty = AccessLevelHelper.GetAccessLevel(modifiers);
+            this.elementPriority = checks.ElementType ? TypeMemberOrder.IndexOf(type) : 0;
+            this.modifierFlags = GetModifierFlags(modifiers, checks);
+            if (checks.AccessLevel)
+            {
+                if ((type == SyntaxKind.ConstructorDeclaration && this.modifierFlags.HasFlag(ModifierFlags.Static))
+                    || (type == SyntaxKind.MethodDeclaration && (member as MethodDeclarationSyntax)?.ExplicitInterfaceSpecifier != null)
+                    || (type == SyntaxKind.PropertyDeclaration && (member as PropertyDeclarationSyntax)?.ExplicitInterfaceSpecifier != null)
+                    || (type == SyntaxKind.IndexerDeclaration && (member as IndexerDeclarationSyntax)?.ExplicitInterfaceSpecifier != null))
```

Comment:
There is no point in checking the type and then redoing the check with `as` and `?.`